main/src/main/java/de/blinkt/openvpn/core/TorProxy.java

```diff
@@ -0,0 +1,316 @@
+
+        executorService.execute(() -> processOutgoingPackets(c));
+
+        executorService.execute(() -> processIncomingPackets(c));
+
+        OrbotHelper orbotHelper = OrbotHelper.get();
+        if (!OrbotHelper.checkTorReceier(c)) {
+            Log.e(TAG, "Orbot does not seem to be installed!");
+        }
+
+        resumeHandler.postDelayed(orbotStatusTimeOutRunnable, ORBOT_TIMEOUT_MS);
+        orbotHelper.addStatusCallback(c, statusCallback);
+
+        orbotHelper.sendOrbotStartAndStatusBroadcast();
+
+        return vpnSide;
+    }
+
+    public static void close() {
+        IPtProxy.stopSocks();
+        if (executorService != null) {
+            executorService.shutdownNow();
+            executorService = null;
+        }
+        if (pipe != null) {
+            for (ParcelFileDescriptor pfd : pipe) {
+                try {
+                    pfd.close();
+                } catch (Exception ignored) {
+
+                }
+            }
+            pipe = null;
+        }
+        if (vpnPfd != null) {
+            try {
+                vpnPfd.close();
+            } catch (Exception ignored) {
+
+            }
+            vpnPfd = null;
+        }
+        if (outputDeviceStream != null) {
+            try {
+                outputDeviceStream.close();
+            } catch (Exception ignored) {
+
+            }
+            outputDeviceStream = null;
+        }
+        if (inputDeviceStream != null) {
+            try {
+                inputDeviceStream.close();
+            } catch (Exception ignored) {
+
+            }
+            inputDeviceStream = null;
+        }
+        if (inputOpenVPNStream != null) {
+            try {
+                inputOpenVPNStream.close();
+            } catch (Exception ignored) {
+
+            }
+            inputOpenVPNStream = null;
+        }
+        if (outputOpenVPNStream != null) {
+            try {
+                outputOpenVPNStream.close();
+            } catch (Exception ignored) {
+
+            }
+            outputOpenVPNStream = null;
+        }
+        isTorConnected = false;
+        pFlow = null;
+        resumeHandler = null;
+    }
+
+    private static void processIncomingPackets(Context c) {
+        ConnectivityManager cm = (ConnectivityManager) c.getSystemService(CONNECTIVITY_SERVICE);
+        PackageManager pm = c.getPackageManager();
+        byte[] buf = new byte[16384];
+        while (!Thread.interrupted()) {
+            try {
+                int read = inputOpenVPNStream.read(buf);
+                if (read > 0) {
+                    int protocol = getProtocol(buf);
+                    if (protocol == 6 || protocol == 17) {
+                        InetSocketAddress sourceAddress = getSourceAddress(buf, protocol);
+                        InetSocketAddress destinationAddress = getDestinationAddress(buf, protocol);
+                        int uid = cm.getConnectionOwnerUid(protocol, destinationAddress, sourceAddress);
```

Comment:
Have you any idea how expensive this call is? It is called for every single VPN packet and is not cached in any way. This might be a considerable slowdown.